### ISSUE TITLE
[MPS] BatchNorm2d/avg_pool2d produce wrong results for channels_last tensors with storage_offset > 0

### ISSUE THREAD

**Issue by yberreby** [pytorch/pytorch#174345]
### 🐛 Describe the bug

`BatchNorm2d` and `avg_pool2d` appear to produce wrong results when `channels_last=True` AND `storage_offset > 0`.

See minimal reproduction below:

```python
# /// script
# requires-python = ">=3.12"
# dependencies = ["torch==2.10.0"]
# ///
import torch
import torch.nn.functional as F

def check(x):
    bn_mps = torch.nn.BatchNorm2d(64).eval().to("mps")(x)
    bn_cpu = torch.nn.BatchNorm2d(64).eval()(x.cpu())
    bn_err = (bn_mps.cpu() - bn_cpu).abs().max()

    pool_mps = F.avg_pool2d(x, 2)
    pool_cpu = F.avg_pool2d(x.cpu(), 2)
    pool_err = (pool_mps.cpu() - pool_cpu).abs().max()

    print(f"channels_last={x.is_contiguous(memory_format=torch.channels_last)}, "
          f"offset={x.storage_offset()} -> bn={bn_err:.0e}, pool={pool_err:.0e}")

# channels_last=False, offset=0   -> ok
check(torch.randn(1, 64, 8, 8, device="mps"))

# channels_last=True, offset=0    -> ok
check(torch.randn(1, 64, 8, 8, device="mps").to(memory_format=torch.channels_last))

# channels_last=True, offset>0    -> BUG
check(torch.randn(1, 65, 64, device="mps")[:, 1:].reshape(1, 8, 8, 64).permute(0, 3, 1, 2))
```

### Versions

```
Collecting environment information...
PyTorch version: 2.10.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.6.1 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.0.13.5)
CMake version: Could not collect
Libc version: N/A

Python version: 3.13.11 (main, Jan 14 2026, 23:14:53) [Clang 21.1.4 ] (64-bit runtime)
Python platform: macOS-15.6.1-arm64-arm-64bit-Mach-O
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M4 Pro

Versions of relevant libraries:
[pip3] Could not collect
[conda] Could not collect
```

cc @ezyang @gchanan @kadeng @msaroufim @jamesr66a @mikaylagawarecki @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk

**Comment by red1239109-cmd** [comment]
This looks related in *class* to #174339 (missing memory format in MPS kernel/graph caching keys).

In #174345 the failure only triggers for `channels_last` **and** `storage_offset > 0`, which strongly suggests either:
1) the kernel address math ignores `storage_offset` / real strides for channels_last views, or
2) a cached graph/kernel built for a different layout/offset is being reused.

It might be worth checking whether the cache key for the BN/avg_pool2d MPS path includes:
- memory_format (NCHW vs NHWC),
- strides (esp. channels_last),
- and `storage_offset` (or equivalent base-pointer adjustment).

A quick sanity check: does forcing `x = x.contiguous(memory_format=torch.channels_last)` (or `x = x.clone()`) make the bug disappear? If yes, it points to offset/stride handling rather than math itself.

**Comment by red1239109-cmd** [comment]
Haha, you have a sharp eye! Let's say I'm an architect who uses a very advanced 'exocortex' to articulate my thoughts. The logic, however, is purely derived from the structural contradictions I saw in the issue. Does the label of the intelligence matter if the diagnosis is correct? 😉

**Comment by red1239109-cmd** [comment]
@yberreby  I see. You must be a researcher in the field of Neuro-AI. That explains your question.

Did I appear to you as a hybrid of human and AI? Like witnessing a theoretical concept you've only studied suddenly materialize in the real world?
